ambry-clustermap/src/main/java/com/github/ambry/clustermap/CompositeDataNodeConfigSource.java

```diff
@@ -0,0 +1,190 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ */
+
+package com.github.ambry.clustermap;
+
+import com.github.ambry.utils.Time;
+import com.github.ambry.utils.Utils;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * An implementation of {@link DataNodeConfigSource} that combines two sources. The "primary" acts as the source of
+ * truth for users of this class, but results obtained are compared against the "secondary" source. This class can be
+ * useful for safe migrations between different backing stores.
+ */
+public class CompositeDataNodeConfigSource implements DataNodeConfigSource {
+  private static final Logger LOGGER = LoggerFactory.getLogger(CompositeDataNodeConfigSource.class);
+  static final long CHECK_PERIOD_MS = TimeUnit.MINUTES.toMillis(5);
+  static final int MAX_PERIODS_WITH_INCONSISTENCY = 3;
+  private final DataNodeConfigSource primarySource;
+  private final DataNodeConfigSource secondarySource;
+  private final DataNodeConfigSourceMetrics metrics;
+  private final Time time;
+  private final AtomicBoolean recordingListenersRegistered = new AtomicBoolean(false);
+
+  /**
+   * @param primarySource the source of truth for users of this class.
+   * @param secondarySource all operations are also attempted against this source, but if the results of the operation
+   *                        differ from the primary, inconsistencies are just recorded with logs and metrics.
+   * @param time a {@link Time} instance used for consistency verification.
+   * @param metrics a {@link DataNodeConfigSourceMetrics} instance to use.
+   */
+  public CompositeDataNodeConfigSource(DataNodeConfigSource primarySource, DataNodeConfigSource secondarySource,
+      Time time, DataNodeConfigSourceMetrics metrics) {
+    this.primarySource = primarySource;
+    this.secondarySource = secondarySource;
+    this.time = time;
+    this.metrics = metrics;
+  }
+
+  @Override
+  public void addDataNodeConfigChangeListener(DataNodeConfigChangeListener userListener) throws Exception {
+    if (recordingListenersRegistered.compareAndSet(false, true)) {
+      RecordingListener primaryListener = new RecordingListener(userListener);
+      RecordingListener secondaryListener = new RecordingListener(null);
+      primarySource.addDataNodeConfigChangeListener(primaryListener);
+      try {
+        secondarySource.addDataNodeConfigChangeListener(secondaryListener);
+      } catch (Exception e) {
+        LOGGER.error("Error from secondarySource.addDataNodeConfigChangeListener({})", userListener, e);
+      }
+      Utils.newThread("DataNodeConfigConsistencyChecker",
+          new ConsistencyChecker(time, metrics, primaryListener, secondaryListener), false).start();
```

Comment:
That is true. If you prefer, I could not start the thread and just introduce a new metric that fires once if attaching the secondary listener fails instead of letting the consistencychecker run and fail continuously.